exonum-java-proofs/src/main/java/com/exonum/binding/storage/proofs/map/flat/UncheckedFlatMapProof.java

```diff
@@ -0,0 +1,269 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * An unchecked flat map proof, which does not include any intermediate nodes.
+ */
+public class UncheckedFlatMapProof implements UncheckedMapProof {
+
+  private static final HashFunction HASH_FUNCTION = Hashing.defaultHashFunction();
+
+  private final List<MapProofEntry> proof;
+
+  private final List<MapEntry> entries;
+
+  private final List<byte[]> missingKeys;
+
+  UncheckedFlatMapProof(
+      List<MapProofEntry> proof,
+      List<MapEntry> entries,
+      List<byte[]> missingKeys) {
+    this.proof = proof;
+    this.entries = entries;
+    this.missingKeys = missingKeys;
+  }
+
+  @SuppressWarnings("unused") // Native API
+  static UncheckedFlatMapProof fromUnsorted(
+      MapProofEntry[] proofList,
+      MapEntry[] entries,
+      byte[][] missingKeys) {
+    List<MapProofEntry> proof = Arrays.asList(proofList);
+    List<MapEntry> entriesList = Arrays.asList(entries);
+    List<byte[]> missingKeysList = Arrays.asList(missingKeys);
+    return new UncheckedFlatMapProof(proof, entriesList, missingKeysList);
+  }
+
+  @Override
+  public CheckedMapProof check() {
+    ProofStatus orderCheckResult = orderCheck();
+    if (orderCheckResult != ProofStatus.CORRECT) {
+      return CheckedFlatMapProof.invalid(orderCheckResult);
+    }
+    if (prefixesIncluded()) {
+      return CheckedFlatMapProof.invalid(ProofStatus.INVALID_STRUCTURE);
+    }
+    if (isEmptyProof()) {
+      return checkEmptyProof();
+    } else if (isSingletonProof()) {
+      return checkSingletonProof();
+    } else {
+      return checkProof();
+    }
+  }
+
+  /**
+   * Checks that all entries in the proof are in the valid order.
+   *
+   * <p>The keys must be in ascending order as defined by
+   * the {@linkplain DbKey#compareTo(DbKey) comparator}; there must not be duplicates.
+   *
+   * @return {@code ProofStatus.CORRECT} if every following key is greater than the previous
+   *         {@code ProofStatus.INVALID_ORDER} if any following key key is lesser than the previous
+   *         {@code ProofStatus.DUPLICATE_PATH} if there are two equal keys
+   * @see DbKey#compareTo(DbKey)
+   */
+  private ProofStatus orderCheck() {
+    for (int i = 1; i < proof.size(); i++) {
+      DbKey key = proof.get(i - 1).getDbKey();
+      DbKey nextKey = proof.get(i).getDbKey();
+      int comparisonResult = nextKey.compareTo(key);
+      if (comparisonResult < 0) {
+        return ProofStatus.INVALID_ORDER;
+      } else if (comparisonResult == 0) {
+        return ProofStatus.DUPLICATE_PATH;
+      }
+    }
+    return ProofStatus.CORRECT;
+  }
+
+  /**
+   * Check if any entry has a prefix among the paths in the proof entries. Both found and absent
+   * keys are checked.
+   */
+  private boolean prefixesIncluded() {
```

Comment:
I'm not sure this is the complete check; we should also check that there are no prefixes of _proof keys_ in other proof keys, too. In the Rust counterpart [this check is performed](https://github.com/exonum/exonum/blob/35617bda142b2d44bf8ea8de753403534e7a9ab3/exonum/src/storage/proof_map_index/proof.rs#L489-L496) concurrently with checking ordering of proof keys. (Technically speaking, skipping this check can lead to an exception when calculating the index Merkle root hash - at least it does in the Rust version - but it may not be desired behavior.)